game/glshader.cc

```diff
@@ -168,6 +168,8 @@ Shader& Shader::link() {
 		throw std::runtime_error("Something went wrong linking the shader program.");
 	}
 	ec.check("glLinkProgram");
+	for (ShaderObjects::const_iterator it = shader_ids.begin(); it != shader_ids.end(); ++it)
+		glDetachShader(program, *it);
```

Comment:
Maybe this could be a for-each loop instead of iterator based? Or ideally, use RAII to detach them instead of manual deletion.

Reply:
This is not deletion, this needs to be done prior to deletion, apparently, deletion is done in destruction of the class. That said, the doc seem to indicate that deletion could happen here as well, as shaders are said to be useless when progam (shaded) is linked.
Do I perform deletion there as well?

Reply:
Sounds like both should be handled in the destructor together, and this loop could be replaced with something like `shader_ids.clear()`.

Reply:
wel that's not that simple, shaders here are just int8, so they don't have destructors. Not sure creating a wrapper class here is really useful.

Reply:
In fact, I believe that the code still leaks memory when linking the program fails, because then an exception is thrown and this cleanup loop is not executed. C++ doesn't have try-finally because you are supposed to use RAII for everything.

Attach/Detach is a clear pair that warrants for a simple RAII wrapper and a separate std::vector inside this function to hold the RAII wrappers.

Ideally, CreateShader/DeleteShader (current code leaks on shader compile error) and LinkProgram/DeleteProgram would be similarly wrapped (possibly using move semantics), but perhaps all that is good to be left for another PR.

Reply:
what about moving the detach just after the `glLinkProgram(program);`? I agree that RAII is the 'normal' way to handle resources, but here, the API is C and calls are within 3 lines of code, not sure adding 20 lines of RAII wrappers would really make the code easier to read... Let me know, if you prefer the RAII, let's open a task for it.

Reply:
That should do.